### PR TITLE
Forgot password and Reset password feature

### PR DESCRIPTION
**Pull request by iamrahulrnair**
The workflow includes:-

- user clicks on forgot password button
- will be prompted to enter the email of user
- if the email is valid, creates a JWT token ,valid for 10 minutes and send to the mail id **(this need to be tested since i don't have the nodemailer acc id and password. testing will be simple just uncomment the transporter segment of code and raise a forgot password )** 
- user can click on that link and reroutes to front-end reset password page,
- user types password and confirm password then server validates by checking the jwt expiry and updates the password for user.
- user cannot raise another request with the same token since that will be security vulnerability.

lemme know if i am missing anything,
 thanks :-)

### CODE REVIEW

File: server/controllers/pwd.controllers.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  var mailOptions = {
+    from: process.env.OUR_EMAIL,
+    to: user.email,
+    cc: 'codedeeper.work@gmail.com',
+    subject: 'Codedeeper password reset Token',
+    html: `<p>Hello , ${user.name} here is your token link. <a href="${resetURL}">${resetURL}</a> to reset the password</p>`,
+    text: `<p>Hello , ${user.name} here is your token link. <a href="${resetURL}">${resetURL}</a> to reset the password</p>`,
+  };
+
+  // transporter.sendMail(mailOptions, function (error, info) {
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:39]: You can remove this comment, I had tested it and it works fine.

File: server/controllers/pwd.controllers.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    },
+    from: 'codesync.live',
+  });
+
+  var mailOptions = {
+    from: process.env.OUR_EMAIL,
+    to: user.email,
+    cc: 'codedeeper.work@gmail.com',
+    subject: 'Codedeeper password reset Token',
+    html: `<p>Hello , ${user.name} here is your token link. <a href="${resetURL}">${resetURL}</a> to reset the password</p>`,
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:35]: You should put it like this
```
    html: `<p>Hello , ${user.name} here is your token link. <a href="${resetURL}">Click Here</a> to reset the password</p>`,
    text: `<p>Hello , ${user.name} here is your token link. <a href="${resetURL}">Click Here</a> to reset the password</p>`,
```

File: server/controllers/pwd.controllers.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
+const asyncHandler = require('express-async-handler');
+const User = require('../models/user.modal');
+const nodemailer = require('nodemailer');
+const jwt = require('jsonwebtoken');
+const bcrypt = require('bcryptjs');
+
+module.exports.forgotPwdController = asyncHandler(async (req, res, next) => {
+  const { email } = req.body;
+  const user = await User.findOne({ email });
+
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:10]: Before generating token we should check lastupdate date&time which should be more than 10 min, if not then send 202 status code with the message "You can do reset password after 10 Min".

File: server/controllers/pwd.controllers.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  if (!user) {
+    res.status(400).send({ status: 'user not found' });
+    throw new Error('User not found');
+  }
+
+  const token = await user.createPWDresetToken();
+  const resetURL = `${req.protocol}://${req.get(
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:17]: //TODO: Here URL should be frontend, You can get from .env file name as FRONTEND_URL

File: server/controllers/pwd.controllers.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  //         resetURL,
+  //       });
+  //     } else {
+  //       res.status(200);
+  //     }
+  //   }
+  // });
+
+  /**
+   * @TODO: check nodemailer authentication and
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:59]: Remove this comments too

File: server/controllers/pwd.controllers.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   * test this in development before building for production
+   */
+  res.status(200).send({
+    resetURL,
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:63]: You should only send messages like ```Password reset link to send via mail``` etc. No need to return the payload here.

File: server/controllers/pwd.controllers.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      throw new Error(
+        'Please raise a new Reset Request,this token was used earlier'
+      );
+
+    const newPassword = await bcrypt.hash(password, 12);
+    user.set({ password: newPassword, passwordChangedAt: new Date() });
+    await user.save();
+
+    return res.send({
+      payload,
```
Comment (codewithvk) [inline server/controllers/pwd.controllers.js:90]: No need to return anything here, just write a message!

File: server/server.js [hunk 8]
```diff
@@ -161,6 +163,12 @@ app.post('/sendMail', (req, res) => {
 //   })
 
 // }
+
+// app.use("*",(error,req,res,next)=>{
```
Comment (codewithvk) [inline server/server.js:167]: Remove it

File: server/routes/pwd.route.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+const {
+  forgotPwdController,
+  resetPasswordController,
+} = require('../controllers/pwd.controllers');
+
+const router = express.Router();
+
+router.get('/_reset_password/:JWT_TOKEN', resetPasswordController);
+router.post('/_forgot_password', forgotPwdController);
```
Comment (codewithvk) [inline server/routes/pwd.route.js:10]: This route should be authenticate, you can look ```authMiddleware```

File: server/controllers/pwd.controllers.js (hunk 3 above)
Comment (codewithvk) [reply server/controllers/pwd.controllers.js:10]: Also in the past user doesn't have the last update field so take care of it.

**Review by codewithvk — CHANGES_REQUESTED** [review]
You are doing great work. Please check the above changes!

File: server/routes/pwd.route.js (hunk 9 above)
Comment (iamrahulrnair) [reply server/routes/pwd.route.js:10]: if the user is authenticated, they would obviously knew the password and does not require _reset_password right?i mean they can't access the route right? 🤔, if i am missing something please inform me the flow, and what  role authmiddleware plays here, thanks

Comment (codewithvk) [reply server/routes/pwd.route.js:10]: Yah ,sorry my mistake :/

File: server/controllers/pwd.controllers.js (hunk 3 above)
Comment (iamrahulrnair) [reply server/controllers/pwd.controllers.js:10]: > Before generating token we should check lastupdate date&tim

This i have done ,could u review line72

Comment (codewithvk) [reply server/controllers/pwd.controllers.js:10]: No, I am talking about while generating tokens. SO that the user can't spam it multiple times.

Comment (iamrahulrnair) [reply server/controllers/pwd.controllers.js:10]: Yeah ok so if a user clicks forgot password, he can't raise another token for 10 mins right?

Comment (codewithvk) [reply server/controllers/pwd.controllers.js:10]: Yes

Comment (iamrahulrnair) [reply server/controllers/pwd.controllers.js:10]: Hey done, added rate limiter :-)

**Review by codewithvk — APPROVED** [review]
LGTM 👍🏻

**Comment by codewithvk** [comment]
@iamrahulrnair Thank you for contributing to our project. Looking forward to work together :)

**Pull request merged**